### PR TITLE
Enable parallel64 bit x unit

### PR DESCRIPTION
**Pull request by agocke**
Try the pull request again. :)

### CODE REVIEW

File: BuildAndTest.proj [hunk 1]
```diff
@@ -43,9 +38,8 @@
   <Target Name="Test">
 
     <ItemGroup Condition="'$(CIBuild)' == ''">
-      <TestAssemblies 
-        Include="Binaries\$(Configuration)\**\*.UnitTests*.dll" 
-        Exclude="Binaries\Roslyn.Compilers.NativeClient.UnitTests.dll" />
+      <TestAssemblies Include="Binaries\$(Configuration)\**\*.UnitTests*.dll"
+                      Exclude="Binaries\$(Configuration)\**\Roslyn*Services*UnitTests*dll"/>
```
Comment (tmeschter) [inline BuildAndTest.proj]: Why aren't we running tests in the Roslyn*Services*UnitTests*dll assemblies?

Comment (agocke) [inline BuildAndTest.proj]: They deadlock. Issue #204

File: src/Compilers/VisualBasic/Test/Semantic/Binding/BindingObjectInitializerTests.vb [hunk 2]
```diff
@@ -1091,7 +1091,7 @@ End Class
         End Sub
 
         <Fact(), WorkItem(788522, "DevDiv")>
-        Sub ObjectInitializerNoStackOverflowFor250LevelsOfNesting()
+        Sub ObjectInitializerNoStackOverflowFor150LevelsOfNesting()
```
Comment (tmeschter) [inline src/Compilers/VisualBasic/Test/Semantic/Binding/BindingObjectInitializerTests.vb]: So it's OK to overflow on 250 levels of nesting now?

File: src/Interactive/HostTest/InteractiveHostTest.csproj [hunk 3]
```diff
@@ -101,8 +101,10 @@
     </ProjectReference>
   </ItemGroup>
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|x86' ">
+    <PlatformTarget>AnyCPU</PlatformTarget>
```
Comment (tmeschter) [inline src/Interactive/HostTest/InteractiveHostTest.csproj]: This seems suspicious. Shouldn't the solution files just be updated to build this as AnyCPU instead of x86?

File: src/Compilers/VisualBasic/Test/Semantic/Binding/BindingObjectInitializerTests.vb (hunk 2 above)
Comment (agocke) [inline src/Compilers/VisualBasic/Test/Semantic/Binding/BindingObjectInitializerTests.vb]: I wouldn't say it's OK, but it's what happens...

It's unlikely we can do much here -- the parsing is inherently recursive and 64-bit frames are larger, so we overflow sooner.

File: src/Test/Utilities/CLRHelpers.cs [hunk 4]
```diff
@@ -11,6 +11,7 @@
 using Microsoft.CodeAnalysis.Test.Utilities.ComTypes;
 using Microsoft.CodeAnalysis.Text;
 using Roslyn.Test.Utilities;
+using System.Threading;
```
Comment (tmeschter) [inline src/Test/Utilities/CLRHelpers.cs]: `using System.*` directives should go first.

File: src/Test/Utilities/CLRHelpers.cs [hunk 5]
```diff
@@ -64,12 +65,14 @@ public static string[] PeVerify(string filePath)
             return PeVerify(File.ReadAllBytes(filePath), AppDomain.CurrentDomain.Id, filePath);
         }
 
-        private static readonly object Guard = new object();
+        private static readonly Mutex s_guard = new Mutex(false, new Guid().ToString());
```
Comment (tmeschter) [inline src/Test/Utilities/CLRHelpers.cs]: A comment about why a Mutex is preferable to a lock object would be useful--it isn't at all clear to me why this change was made.

Comment (agocke) [inline src/Test/Utilities/CLRHelpers.cs]: Good point, I'll add one. It's because locking doesn't extend past app domain.

Comment (tmeschter) [inline src/Test/Utilities/CLRHelpers.cs]: Is `s_guard` shared between AppDomains, or does each AppDomain get a different copy? If it's the latter then how does that help?

Comment (agocke) [inline src/Test/Utilities/CLRHelpers.cs]: Hmm -- you're right. This is probably passing accidentally because locking a mutex takes so long.

However, locking with the same guid should work. I'll hardcode the GUID into the file.

Comment (tmat) [inline src/Test/Utilities/CLRHelpers.cs]: This creates system-wide named mutex, right? I don't think we want be taking system-wide locks.

Comment (agocke) [inline src/Test/Utilities/CLRHelpers.cs]: I couldn't find a synchronization routine more course grained than named mutex that locks across app-domain. Suggestions?

Comment (tmeschter) [inline src/Test/Utilities/CLRHelpers.cs]: Yeah, that'll work--no worse than any other hard-coded string.

What resource is the Mutex protecting?

Comment (agocke) [inline src/Test/Utilities/CLRHelpers.cs]: PeVerify does an unsafe call into the CLR to do verification. This routine isn't thread safe and crosses app domain (since it's in the CLR).

File: src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinueTests.vb [hunk 6]
```diff
@@ -3325,7 +3326,7 @@ End Namespace
                                                                      testData0.GetMethodData("M.B.M").EncDebugInfoProvider)
                 Dim method0 = compilation0.GetMember(Of MethodSymbol)("M.B.M")
                 Dim reader0 = md0.MetadataReader
-                CheckNames(reader0, reader0.GetTypeDefNames(), "<Module>", "VB$AnonymousType_1`2", "VB$AnonymousType_0`2", "VB$AnonymousType_2`1", "A", "B")
+                CheckNames(reader0, reader0.GetTypeDefNames(), "<Module>", "VB$AnonymousType_1`2", "VB$AnonymousType_2`1", "VB$AnonymousType_0`2", "A", "B")
```
Comment (tmat) [inline src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinueTests.vb]: Unless you fixed the compiler to deterministically enumerate anonymous types this will break. Currently the compiler enumerates a dictionary, so the order depends on the has code algorithm used by the dictionary, which is presumably different between x86 and x64. There is an issue filed to address this problem.

Comment (agocke) [inline src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinueTests.vb]: Yes, I field the issue. :)

In the meantime we can 1) disable the test or 2) pick an architecture that it will succeed on and only run tests in that arch. That arch is amd64. :)

File: src/Test/Utilities/CLRHelpers.cs (hunk 5 above)
Comment (tmat) [inline src/Test/Utilities/CLRHelpers.cs]: Maybe use the Process ID as the name?

Comment (agocke) [inline src/Test/Utilities/CLRHelpers.cs]: Sure.

File: src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinueTests.vb (hunk 6 above)
Comment (tmat) [inline src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinueTests.vb]: OK. Alternatively we could also fix the bug. It shouldn't be that hard. ;)

File: BuildAndTest.proj (hunk 1 above)
Comment (jaredpar) [inline BuildAndTest.proj]: Disabling that much testing seems dangerous.  Why can't this be fixed before hand?

Comment (agocke) [inline BuildAndTest.proj]: Jason's already fixed it, I think. We just need to integrate.

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs [hunk 7]
```diff
@@ -3328,7 +3328,7 @@ .locals init (object V_0)
         /// containing the current frame, so new instances can
         /// be used interchangeably with existing instances.
         /// </summary>
-        [Fact(Skip = "error CS0019: Operator '==' cannot be applied to operands of type ...")]
+        [Fact(Skip = "Shows 0019: Operator '==' cannot be applied to operands of type ...")]
```
Comment (tmeschter) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs]: What's the work item associated with this?

**Comment by tmeschter** [comment]
:+1:

File: BuildAndTest.proj [hunk 8]
```diff
@@ -70,15 +64,15 @@
...
-      <TestAssemblies Include="Binaries\Debug\Roslyn.Services.VisualBasic.UnitTests.dll" />
+      <TestAssemblies Include="Binaries\Debug\Roslyn.Services.VisualBasic.UnitTests.dll" /> -->
     </ItemGroup>
 
-    <xunit
-      Assemblies="@(TestAssemblies)"
-      Html="UnitTestResults.html"
-      ShadowCopy="false" />
+    <exec
+      Command="packages\xunit.runners.2.0.0-rc1-build2826\tools\xunit.console.exe @(TestAssemblies, ' ') -noshadow -parallel all" />
```
Comment (KevinH-MS) [inline BuildAndTest.proj]: Perhaps this path should be made into a global property in one of the common targets files (the path to xUnit)...  Unrelated to this change, but I notice that even though we're using the 2.0 runner, we are still referencing xunit.1.9.2 from all of our test projects (perhaps this <Reference> should be in a common targets file as well).

**Comment by jaredpar** [comment]
I chatted with @agocke about this and the plan is to break it up into multiple commits that individually address the following:

1. Get the developer loop for testing back down to test.cmd levels we were familiar with.  I'm pushing to get that done today. 
2. Fix the issues that occur when running our unit tests for portable level components in 64 bit. 
3. Fix the issues that occur when running our unit tests for portable level components in parallel. 
4. Using xunit to control our test level parallelization
